pyClient/zeth/signing.py

```diff
@@ -0,0 +1,145 @@
+    y = FQ(
+        int(bytes(urandom(key_size_byte)).hex(), 16) % constants.ZETH_PRIME)
+    X = ec.multiply(ec.G1, x.n)
+    Y = ec.multiply(ec.G1, y.n)
+    return SchnorrKeyPair(x, y, X, Y)
+
+def gen_vk(sk: SchnorrSigningKey) -> SchnorrVerificationKey:
+    """
+    Generate an OT-Schnorr verification key from an OT-Schnorr signing key
+    """
+    x = sk.psk
+    y, _ = sk.ssk
+    X = ec.multiply(ec.G1, x.n)
+    Y = ec.multiply(ec.G1, y.n)
+    return SchnorrVerificationKey(X, Y)
+
+
+def encode_group_element(group_el: G1) -> bytes:
+    """
+    Encode a group element into a byte string
+    We assume here the group prime $p$ is written in less than 256 bits
+    to conform with Ethereum bytes32 type.
+    """
+    res = encode_abi(
+        ["bytes32", "bytes32"],
+        [
+            bytes.fromhex("{0:0>64X}".format(int(group_el[0]))),
+            bytes.fromhex("{0:0>64X}".format(int(group_el[1])))
+        ]
+    )
+    return res
+
+
+def encode_vk(vk: SchnorrVerificationKey) -> bytes:
+    """
+    Encode a verification key as a byte string
+    We assume here the group prime $p$ is written in less than 256 bits
+    to conform with Ethereum bytes32 type
+    """
+    vk_byte = encode_group_element(vk.ppk)
+    vk_byte += encode_group_element(vk.spk)
+    return vk_byte
+
+
+def sign(
+        sk: SchnorrSigningKey,
+        m: str) -> int:
+    """
+    Generate a Schnorr signature on a message m.
+    We assume here that the message is an hexadecimal string written in
+    less than 256 bits to conform with Ethereum bytes32 type.
+    """
+
+    # Encode and hash the verifying key and input hashes
+    challenge_to_hash = encode_group_element(sk.ssk[1]) + encode_single("bytes32", bytes.fromhex(m))
+
+    # Convert the hex digest into a field element
+    challenge = int(sha256(challenge_to_hash).hexdigest(), 16) % constants.ZETH_PRIME
+
+    # Compute the signature sigma
+    sigma = (sk.ssk[0].n + challenge * sk.psk.n) % constants.ZETH_PRIME
+
+    return sigma
+
+def verify(
+    vk: SchnorrVerificationKey,
+    m: str,
+    sigma: int) -> bool:
+    """
+    Return true if the signature sigma is valid on message m and vk.
+    We assume here that the message is an hexadecimal string written in
+    less than 256 bits to conform with Ethereum bytes32 type.
+    """
+    # Encode and hash the verifying key and input hashes
+    challenge_to_hash = encode_group_element(vk.spk) + encode_single("bytes32", bytes.fromhex(m))
+
+    challenge = int(sha256(challenge_to_hash).hexdigest(), 16) % constants.ZETH_PRIME
+
+    left_part = ec.multiply(ec.G1, FQ(sigma).n)
+    right_part = ec.add(vk.spk, ec.multiply(vk.ppk, FQ(challenge).n))
+
+    return ec.eq(left_part, right_part)
+
+def test_all() -> None:
+    """
+    Unit test on the package
+    """
+    m = urandom(32).hex()
+    keypair = key_gen()
+    sigma = sign(keypair.sk, m)
+    assert verify(keypair.vk, m, sigma)
```

Comment:
Generate a test that should fail, by replacing the vk by some garbage, and make sure it fails by asserting the result of verify to be false

Reply:
updated